File: implement_queue_using_stack.cpp
//push efficient pop complex



class MyQueue {
public:
    stack<int> st ; 
    stack<int> helper ;
    MyQueue() {
        
    }
    
    void push(int x) {
        st.push(x);
        
    }
    
    int pop() { // remove at Bottom 
    while( st.size() > 0){
        helper.push(st.top());
        st.pop();

    }
    int x = helper.top();
    helper.pop();
    while( helper.size() > 0){
        st.push(helper.top());
        helper.pop();

    }
    return x;
    }
    
    int peek() {  // retrieve 

    while( st.size() > 0){
        helper.push(st.top());
        st.pop();

    }
    int x = helper.top();
    
    while( helper.size() > 0){
        st.push(helper.top());
        helper.pop();

    }
    return x;
    }


         
    
    
    bool empty() {
        if(st.size() == 0 ) return true ;
        else return false ;
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */




// pop efficeint push complex
void push(int x ){
    // push at bottom 

    if ( st.size() ==0) {
        st.push(x) ;
        return ;
    }
    else{
        while ( st.size() != 0){
            helper.push(st.top())
                st.pop();
        }
        st.push(x);
        while(helper.size() >0 ){

            st.push(helper.top());
            helper.pop();
        }
        
    }
}


void pop( ){
    int x = st.top();
    st.pop(x);
    return x;

}

File: implement_queue_using_stack.java
import java.util.Stack;

class MyQueue {
    Stack<Integer> st;
    Stack<Integer> helper;

    // Constructor
    public MyQueue() {
        st = new Stack<>();
        helper = new Stack<>();
    }

    // Insert element into the queue
    public void push(int x) {
        st.push(x);
    }

    // Remove element from the front of the queue
    public int pop() {
        // Move all elements to helper
        while (!st.isEmpty()) {
            helper.push(st.pop());
        }

        // Remove the top element (front of queue)
        int x = helper.pop();

        // Move back elements to st
        while (!helper.isEmpty()) {
            st.push(helper.pop());
        }

        return x;
    }

    // Get the front element
    public int peek() {
        // Move all elements to helper
        while (!st.isEmpty()) {
            helper.push(st.pop());
        }

        // Front element
        int x = helper.peek();

        // Move back elements to st
        while (!helper.isEmpty()) {
            st.push(helper.pop());
        }

        return x;
    }

    // Check if queue is empty
    public boolean empty() {
        return st.isEmpty();
    }
}


// Example usage
class Main {
    public static void main(String[] args) {
        MyQueue obj = new MyQueue();
        obj.push(10);
        obj.push(20);
        obj.push(30);

        System.out.println(obj.pop());  // 10
        System.out.println(obj.peek()); // 20
        System.out.println(obj.empty()); // false
    }
}
